Sources/StreamChatUI/ChatChannelList/Search/ChatMessageSearchVC.swift

```diff
@@ -0,0 +1,161 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import StreamChat
+import UIKit
+
+/// The view controller responsible to search messages.
+open class ChatMessageSearchVC: ChatChannelListVC, UISearchResultsUpdating, ChatMessageSearchControllerDelegate {
```

Comment:
This looks quite similar to the other subclass. What about a common class that hold most of the logic, contains also some logic that should not be in ChatChannelListVC, and the final concrete implementations subclass from this intermediate class?

Reply:
The reason it is two separate classes for search is that otherwise, it will require a lot of if-statements which IMO will make it harder to maintain but also harder to customise for the customers. It seems to me that separating the logic into two classes can scale better, and it is also more flexible. 

I played with multiple approaches, and this one was the approach that worked best for us and for the customers. We need to keep in mind always in terms of customizability and reusability.

Because this approach is flexible enough, it means if we want to reuse some of the logic, we can in the future. Right now, it could be a premature decision to try to remove duplication, but let me know if you have a different opinion 👍

Reply:
What I meant by my comment is not to have a common class for both use cases, but instead having an intermediate one that holds the common logic

ChatChannelListVC -> ChatChannelSearchVC (internal) -> ChatMessageSearchVC

That way you have all the common logic in ChatChannelSearchVC. If there is any issue in search, we solve it once and for all its subclasses, which makes it easier to maintain. ChatMessageSearchVC can then have only customization options that apply to the particular use case.

Reply:
#### So you mean something like the following:

```mermaid
classDiagram
    ChatChannelListSearchVC <|-- ChatChannelSearchVC
    ChatChannelListVC <|-- ChatChannelListSearchVC
    ChatChannelListSearchVC <|-- ChatMessageSearchVC
    class ChatChannelSearchVC{
    }
    class ChatChannelListSearchVC{
    }
    class ChatMessageSearchVC{
    }
```

Code here:

```swift
open class ChatChannelListSearchVC: ChatChannelListVC, UISearchResultsUpdating {
    /// The component responsible to debounce search requests.
    public var debouncer = Debouncer(0.3, queue: .main)

    /// The current active search text.
    public var currentSearchText: String = ""

    /// A component responsible to handle when to load new channels.
    private lazy var viewPaginationHandler: ViewPaginationHandling = {
        ScrollViewPaginationHandler(scrollView: collectionView)
    }()

    override open func setUpLayout() {
        super.setUpLayout()

        navigationItem.searchController?.searchBar.placeholder = "Why"

        view.embed(emptyView)
        emptyView.isHidden = true
        emptyView.actionButton.removeFromSuperview()
        emptyView.titleLabel.isHidden = true
    }

    override open func setUp() {
        collectionView.register(
            components.channelCell.self,
            forCellWithReuseIdentifier: collectionViewCellReuseIdentifier
        )

        collectionView.register(
            components.channelCellSeparator,
            forSupplementaryViewOfKind: ListCollectionViewLayout.separatorKind,
            withReuseIdentifier: separatorReuseIdentifier
        )

        collectionView.dataSource = self
        collectionView.delegate = self

        viewPaginationHandler.bottomThreshold = 800
        viewPaginationHandler.onNewBottomPage = { [weak self] in
            self?.loadMoreChannels()
        }
    }

    override open func setUpAppearance() {
        super.setUpAppearance()

        emptyView.iconView.image = appearance.images.emptySearch
    }

    open func updateSearchResults(for searchController: UISearchController) {
        guard let text = searchController.searchBar.text, !text.isEmpty, text != currentSearchText else {
            return
        }

        currentSearchText = text

        debouncer.execute { [weak self] in
            self?.updateSearchResults(with: text)
        }
    }

    // swiftlint:disable unavailable_function
    open func updateSearchResults(with text: String) {
        fatalError("This function should be implemented by a subclass.")
    }

    // swiftlint:disable unavailable_function
    open func loadMoreSearchResults() {
        fatalError("This function should be implemented by a subclass.")
    }

    open var hasEmptyResults: Bool {
        fatalError("This function should be implemented by a subclass.")
    }

    override open func controller(_ controller: DataController, didChangeState state: DataController.State) {
        switch state {
        case .initialized, .localDataFetched:
            if hasEmptyResults {
                loadingIndicator.startAnimating()
            } else {
                loadingIndicator.stopAnimating()
            }
        case .remoteDataFetched:
            loadingIndicator.stopAnimating()
            emptyView.subtitleLabel.text = L10n.ChannelList.Search.Empty.subtitle("\"\(currentSearchText)\"")
            emptyView.isHidden = !hasEmptyResults
        default:
            loadingIndicator.stopAnimating()
        }
    }

    deinit {
        debouncer.invalidate()
    }
}

/// The view controller responsible to search channels.
open class ChatChannelSearchVC: ChatChannelListSearchVC {
    /// The closure that is triggered whenever a channel is selected from the search result.
    public var didSelectChannel: ((ChatChannel) -> Void)?

    open override var hasEmptyResults: Bool {
        controller.channels.isEmpty
    }

    open override func updateSearchResults(with text: String) {
        guard let currentUserId = controller.client.currentUserId else { return }

        replaceQuery(.init(
            filter: .and([
                .autocomplete(.name, text: text),
                .containMembers(userIds: [currentUserId])
            ])
        ))
    }

    open override func loadMoreSearchResults() {
        loadMoreChannels()
    }

    override open func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        defer {
            collectionView.deselectItem(at: indexPath, animated: true)
        }
        guard let channel = controller.channels[safe: indexPath.row] else { return }
        didSelectChannel?(channel)
    }
}

/// The view controller responsible to search messages.
open class ChatMessageSearchVC: ChatChannelListSearchVC, ChatMessageSearchControllerDelegate {
    /// The `ChatMessageSearchController` instance to perform the messages search.
    public var messageSearchController: ChatMessageSearchController!

    /// The closure that is triggered whenever a message is selected from the search result.
    public var didSelectMessage: ((ChatChannel, ChatMessage) -> Void)?

    /// Component responsible to process an array of `[ListChange<Item>]`'s and apply those changes to a view.
    private lazy var listChangeUpdater: ListChangeUpdater = CollectionViewListChangeUpdater(
        collectionView: collectionView
    )

    private var isPaginatingMessages: Bool = false

    open override var hasEmptyResults: Bool {
        messageSearchController.messages.isEmpty
    }

    override open func setUp() {
        super.setUp()

        messageSearchController.delegate = self
    }

    open override func updateSearchResults(with text: String) {
        messageSearchController.search(text: text)
    }

    open override func loadMoreSearchResults() {
        loadMoreMessages()
    }

    override open func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        messageSearchController.messages.count
    }

    override open func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(with: ChatChannelListCollectionViewCell.self, for: indexPath)
        guard let message = messageSearchController.messages[safe: indexPath.item],
              let cid = message.cid,
              let channel = messageSearchController.dataStore.channel(cid: cid) else {
            return cell
        }

        cell.components = components
        cell.itemView.content = .init(
            channel: channel,
            currentUserId: messageSearchController.client.currentUserId,
            searchedMessage: message
        )

        cell.swipeableView.delegate = self
        cell.swipeableView.indexPath = { [weak cell, weak self] in
            guard let cell = cell else { return nil }
            return self?.collectionView.indexPath(for: cell)
        }

        return cell
    }

    override open func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        defer {
            collectionView.deselectItem(at: indexPath, animated: true)
        }

        guard let message = messageSearchController.messages[safe: indexPath.item],
              let cid = message.cid,
              let channel = messageSearchController.dataStore.channel(cid: cid) else {
            return
        }

        didSelectMessage?(channel, message)
    }

    open func loadMoreMessages() {
        guard !isPaginatingMessages else {
            return
        }
        isPaginatingMessages = true

        messageSearchController.loadNextMessages { [weak self] _ in
            self?.isPaginatingMessages = false
        }
    }

    open func controller(_ controller: ChatMessageSearchController, didChangeMessages changes: [ListChange<ChatMessage>]) {
        listChangeUpdater.performUpdate(with: changes)
    }
}
```

My main issue here is that, as I said before, it could be a bit of premature abstraction, especially because it adds a bit of complexity. Instead of having just 2 classes only, we now have 3. Since the duplication is minimal, it could not be worth the additional code and complexity. But I'm fine with both options. Like I said before as well, if we decide not to go with this additional abstraction, we can always introduce `ChatChannelListSearchVC` in the future. But if we do introduce it now, we can never remove it.